conbench/api/history.py

```diff
@@ -42,10 +53,146 @@ def get(self, benchmark_result_id):
+            in: path
+            schema:
+                type: string
+        tags:
+          - History
+        """
+        # TODO: think about the case where samples if of zero length. Can this
+        # happen? If it can happen: which response would we want to emit to the
+        # HTTP client? An empty array, or something more convenient?
+        try:
+            items = get_history_for_benchmark(benchmark_result_id=benchmark_result_id)
+        except NotFound:
+            self.abort_404_not_found()
+
+        csv_buf = generate_csv_history_for_result(benchmark_result_id, items)
+
+        return send_file(
+            csv_buf,
+            as_attachment=True,
+            # Use a "timeseries" fingerprint here that represents what all data
+            # points in this series have in common: benchmark name, case perm,
+            # hardware, context, repo.
+            download_name=f"conbench-history-{items[0].benchmark_name}-{items[0].ts_fingerprint}.csv",
+            mimetype="text/csv",
+        )
+
+        # # Note: wrap this into an array if there is just 1 sample, for
+        # # consistency (clients expect an array).
+        # jsonbytes: bytes = orjson.dumps(
+        #     [s._dict_for_api_json() for s in samples],
+        #     option=orjson.OPT_INDENT_2,
+        # )
+        # return csv_response_for_byte_sequence(csv_buf.getbuffer(), 200)
+
+
 history_entity_view = HistoryEntityAPI.as_view("history")
+history_download_endpoint = HistoryDownloadAPI.as_view("history-download")
+
+rule(
+    "/history/download/<benchmark_result_id>/",
+    view_func=history_download_endpoint,
+    methods=["GET"],
+)
+
 
 rule(
     "/history/<benchmark_result_id>/",
     view_func=history_entity_view,
     methods=["GET"],
 )
+
+# deduplicate: move to conbench.numstr
+numstr8 = functools.partial(conbench.numstr.numstr, sigfigs=8)
+
+
+def generate_csv_history_for_result(
+    input_result_id: str, items: List[HistorySample]
+) -> BytesIO:
+    """
+    input from `get_history_for_benchmark()`. Intrinsics of that function
+    matter a lot; read its docstring and code. For example, two important
+    aspects about that func:
+    - results for default branch only
+    - order is not guaranteed in returned collection.
+    """
+
+    assert len(items) > 0
+
+    # Note that this might start to have similarities to the dataframe aspects
+    # within entities.history.execute_history_query_get_dataframe() -- that's
+    # expected. let's streamline internal and external interface in the future
+    # and then we can make analysis on these dataframes (easy-ish) testable.
+    df = pd.DataFrame(
+        # Note(jp:): cannot use a generator expression here, len needs
+        # to be known.
+        {
+            "result_id": [i.benchmark_result_id for i in items],
+            "commit_hash": [i.commit_hash for i in items],
+            "svs": [i.svs for i in items],
+            "min": [min(i.data) for i in items],
+        },
```

Comment:
> I suspect we may wan to vary what goes in here over time

absolutely

> we should mark this as experimental and liable to change

absolutely, too. I mean. Everything we do is liable to change, honestly. But this here is an extreme version, yes. 

I have started thinking about treating (and documenting) file contents as implementation detail (and care about stability of corresponding parsing code).

In any case: this is for starters for us, mainly aimed at allowing us to build up a library of test cases. I also think this can be a more widely useful interface in the future. We will see.